fiber-unix/src/scheduler.ml

```diff
@@ -446,3 +418,174 @@ let cancel_timer (timer : timer) =
 let abort t =
   (* TODO proper cleanup *)
   add_events t [ Abort ]
+
+module Process_watcher : sig
+  val init : t -> process_watcher
+
+  (** Register a new running process. *)
+  val register : process_watcher -> process -> unit
+
+  (** Send the following signal to all running processes. *)
+  val killall : process_watcher -> int -> unit
+end = struct
+  module Process_table : sig
+    val add : process_watcher -> process -> unit
+
+    val remove : process_watcher -> pid:Pid.t -> Unix.process_status -> unit
+
+    val running_count : process_watcher -> int
+
+    val iter : process_watcher -> f:(process -> unit) -> unit
+  end = struct
+    let add t job =
+      match Table.find t.table job.pid with
+      | None ->
+        Table.set t.table job.pid (Running job);
+        t.running_count <- t.running_count + 1;
+        if t.running_count = 1 then Condition.signal t.something_is_running
+      | Some (Zombie status) ->
+        Table.remove t.table job.pid;
+        add_events t.process_scheduler [ Job_completed (status, job.ivar) ]
+      | Some (Running _) -> assert false
+
+    let remove t ~pid status =
+      match Table.find t.table pid with
+      | None -> Table.set t.table pid (Zombie status)
+      | Some (Running job) ->
+        t.running_count <- t.running_count - 1;
+        Table.remove t.table pid;
+        add_events t.process_scheduler [ Job_completed (status, job.ivar) ]
```

Comment:
Does the `Job_completed` event do anything special? If not we should probably kill the process when calling `remove`

Reply:
Note that we kill remove on the process after we did `Unix.wait` on it. So it's already dead.